docs/design/design.md

```diff
@@ -0,0 +1,116 @@
+* **Robustness to failure modes**: We expect and gracefully handle failure modes stemming from field-deployment and that MicroShift is just an app on somebody else's OS it cannot control.
+* **Production over Development**: We engineer for production-deployments, not Kubernetes application development environments.
+* **Frugal use of resources**: We are mindful of MicroShift's resources consumption, considering all resources (memory, CPU, storage I/O, network I/O).
+* **No premature resouce optimization**: We do not attempt to squeeze out the last 20% of resouces, e.g. by patching or compressing code, inventing lighter-weight components, etc.
+* **Ease-of-Use**: We make MicroShift intutive and simple to use, even if that means providing fewer features and options.
+* **Prefer well-established mechanisms**: We meet users where they are by letting them use well-established tools and patterns rather than requiring them to learn new ones.
+* **Cheap Control Plane Restarts**: We keep MicroShift restarts cheap (= fast, low resource consumption, low workload impact, etc.) as it is our default model for software and configuration changes.
+* **Alignment with OpenShift**: We reuse OpenShift's code, operational logic, and production chain tools and processes where possible with reasonable effort, unless this would conflict with MicroShift's goals.
+
+
+## Design Decisions
+### Overall Architecture
+* MicroShift is an application deployed onto a running OS, preferably as container on `podman`, managed through `systemd`. As such, it cannot assume any responsiblity or control over the device or OS it runs on, including OS software or configuration updates or typical device management tasks such as configuring host CA certs or host telemetry. 
+* MicroShift runs as a single binary embedding as goroutines only those services strictly necessary to bring up a *minimal Kubernetes/OpenShift control and data plane*. Motivation:
+  * Maximizes reproducibility; cluster will come up fully or not at all.
+  * Does not require external orchestration, for example through operators, and allows for very fast start-up/update times.
+  * Makes it simple to grok as workload for a a Linux admin persona, works well / easier to implement with systemd.
+  * Smaller resource footprint has _not_ been a motivation, it may be a welcome side-effect.
+* MicroShift provides a small, optional set of infrastructure services to support common use cases and reuses OpenShift's container images for these:
+  * openshift-dns, openshift-router, service-ca, local storage provider
+* MicroShift instances (processes) run directly on the host or containerized on Podman. They can take on the roles of Control Plane, Node, or both:
+  * Instances with Control Plane role run etcd and the Kubernetes and OpenShift control plane services. As these services don't require a kubelet, pure Control Plane instances are not nodes in the Kubernetes sense and require fewer system privileges.
```

Comment:
I'm not sure this bullet point is completely correct.

For example, in-cluster k8s-API access by pods. Some CNIs, (and submariner as another example), require running pods on the control nodes too. To make sure that pods can reach (via pod network) the control nodes too. Normally those nodes are tainted to avoid regular user workloads from crawling in.

I guess that as long as we can have the possibility of the control plane running kubelet it should be fine.

Reply:
After reading the other points this requires no action :)